### ISSUE TITLE
Use Illuminate/Schema in installer, instead of raw queries, to support SQL Server, Postgres, and SQLite automatically

### ISSUE THREAD

**Issue by xoorath** [userfrosting/UserFrosting#541]
Hey,

I'm investigating userfrosting for an upcoming project. We would like to use our azure SQL db but I'm not sure how to connect successfully from userfrosting.

A connection would normally be made with the following:
`$connection = new PDO( "sqlsrv:server = tcp:$HOST,$PORT; Database = $DB", $USER, $PASS);`

But I'm unsure how I can tell userfrosting to make a connection with sqlsrv, or include tcp and the port in the hostname? I guess for the tcp+port I would just set `db_host` to include tcp and the port? but I'm still not sure how I could configure userfrosting to use sqlsrv...

Any ideas?

**Comment by alexweissman** [comment]
Eloquent (the database abstraction package we use) definitely supports SQL Server.  The connection is initialized in [initialize.php](https://github.com/userfrosting/UserFrosting/blob/master/userfrosting/initialize.php#L42-L51).  

For more information on `Capsule`, which is the standalone adapter for Eloquent, see https://github.com/illuminate/database/blob/master/README.md.

**Comment by xoorath** [comment]
Hey @alexweissman,

I'm sorry to bug you about this, I'm sure this isn't a userfrosting problem. If you could help point me in the right direction that would be really appreciated though....

I think I must be writing my connection strings wrong, I can't connect using any variation other than _exactly_ the string I posted above, that I got from azure. I'm not sure how to format my connection parameters in userfrosting so I end up with the "sqlsrv: server=tcp:host"<and so on>. 

Do you think I might be missing some configuration step in azure, or if it's a config issue? Any advice is appreciated.... Thanks so much.

**Comment by xoorath** [comment]
Hey,

So I'm making some progress. I added a `db_port` option to the config and linked it into the `initialize.php`, and I also changed the driver in the `initializephp` to `sqlsrv` like our azure database is expecting. So at least I can connect now.

However now we're failing when trying to execute a query. It's complaining that the SQL isn't valid, probably because sql server expects TSQL.

The error I get is (with trimmed paths and comments removed for brevity):
`
Error in vendor\\illuminate\\database\\Connection.php on line 624: SQLSTATE[42000]: [Microsoft][ODBC Driver 11 for SQL Server][SQL Server]Incorrect syntax near the keyword 'IF'. (SQL: CREATE TABLE IF NOT EXISTS `uf_configuration` (
            `id` int(10) unsigned NOT NULL AUTO_INCREMENT,
            `plugin` varchar(50) NOT NULL,
            `name` varchar(150) NOT NULL,
            `value` longtext NOT NULL,
            `description` text NOT NULL,
            PRIMARY KEY (`id`)
        ) ENGINE=InnoDB  DEFAULT CHARSET=utf8 AUTO_INCREMENT=1 ;)

#0 vendor\\illuminate\\database\\Connection.php(580): Illuminate\\Database\\Connection->runQueryCallback('CREATE TABLE IF...', Array, Object(Closure))
#1 vendor\\illuminate\\database\\Connection.php(363): Illuminate\\Database\\Connection->run('CREATE TABLE IF...', Array, Object(Closure))
#2 models\\database\\Database.php(174): Illuminate\\Database\\Connection->statement('CREATE TABLE IF...')
#3 controllers\\InstallController.php(79): UserFrosting\\Database::install()
#4 public\\index.php(343): UserFrosting\\InstallController->pageSetupDB()
#5 [internal function]: {closure}()
#6 vendor\\slim\\slim\\Slim\\Route.php(468): call_user_func_array(Object(Closure), Array)
#7 vendor\\slim\\slim\\Slim\\Slim.php(1357): Slim\\Route->dispatch()
#8 vendor\\slim\\slim\\Slim\\Middleware\\Flash.php(85): Slim\\Slim->call()
#9 vendor\\slim\\slim\\Slim\\Middleware\\MethodOverride.php(92): Slim\\Middleware\\Flash->call()
#10 middleware\\CsrfGuard.php(51): Slim\\Middleware\\MethodOverride->call()
#11 middleware\\usersession\\UserSession.php(27): Slim\\Extras\\Middleware\\CsrfGuard->call()
#12 vendor\\slim\\slim\\Slim\\Slim.php(1302): UserFrosting\\UserSession->call()
#13 public\\index.php(431): Slim\\Slim->run()
#14 {main}
`

So I guess where I'm at now is not knowing if there's a flag to enable wrapping the SQL calls in an execute line that TSQL will understand, or if I can tell SQL Server to accept SQL instead of just TSQL. *OR* If I'm just completely out to lunch and I don't understand the root of the problem.

Again: if you have any idea what could be going on here any input is appreciated.

Thanks @alexweissman

**Comment by alexweissman** [comment]
So, the problem is that the installer (and I think at this point, *only* the installer) is using raw queries instead of Eloquent to set up the initial tables and values.  There are basically two options:

1. You could just create those tables manually in TSQL, instead of using the installer.
2. You could take a crack at migrating the installer over to Eloquent.  If you wanted to look into [creating new tables in Eloquent](https://laravel.com/docs/5.0/schema), and port the installer code over to that, we'd all really appreciate the contribution.  Then we could truly say that UF is database-agnostic!

**Comment by xoorath** [comment]
I would love to do option 2 if this were a personal project, but I do have to work against a time budget here and I'm not sure it will be feasible to do so. I'm not a PHP expert so neither of us would probably be happy with my end product under the time constraint I have.

For option 1: if I do double back to staying with azure and php then I'll likely do this. I might end up moving to a `mean` stack though, where I'm more comfortable. It depends if I can get the approval to get the server I need for that.

Sorry if that's kind of a bummer! Maybe I'll come back to fix the installer at a later date, but I'm not optimistic this will be on my todo list any time soon. I'll change the title of this issue to reflect the real problem here.

**Comment by alexweissman** [comment]
No problem, I completely understand.  MEAN seems to be popular these days, so if you're comfortable with that, I'd say go for it.

Good luck!

**Comment by alexweissman** [comment]
Also, I'm gonna go ahead and leave this open as the canonical issue for a db-agnostic installer.

**Comment by m2jobe** [comment]
Hi @xoorath  were you able to get everthing working?

**Comment by xoorath** [comment]
Hey @m2jobe , nah I moved away from User Frosting.

To my understanding the issue still remains, I don't see any submits against it.

**Comment by Netrilix** [comment]
There's a prototype of it, but we ended up using it was a starting point for v4, rather than merging it into 0.3.1. If there's still demand for it, I can try to work on it soon, but I've been away from coding for a few weeks with a bunch of other projects.

Here's what I had, it only creates a few of the tables so far:
https://gist.github.com/Netrilix/30ea24925ce3be650ef856d2c79c8bb8

**Comment by lcharette** [comment]
This has been implemented by @alexweissman for V4. You can also use that as a starter points (tables did changed for V4). See: https://github.com/userfrosting/UserFrosting/blob/dev/migrations/schema.php

**Comment by alexweissman** [comment]
I think this would be something worthwhile implementing for 0.3.1 as well.  It shouldn't take more than 30 minutes or so to implement and test.

**Comment by alexweissman** [comment]
oops didn't mean to close.

**Comment by m2jobe** [comment]
Hi @lcharette  has it been fully implemented for sqlsrv? I believe I was able to successfully connect as seen here http://imgur.com/a/lGYrc, the tables arent created but I believe I receive that output because it requires tsql, if i were to manually create all those tables, Im assuming everything should work accordingly?

**Comment by m2jobe** [comment]
Where are the uf_authorise etc tables being created in the Userfrosting folder?

**Comment by lcharette** [comment]
Not sure I understand the last question?

**Comment by alexweissman** [comment]
https://github.com/userfrosting/UserFrosting/blob/master/userfrosting/models/database/Database.php#L159-L268

**Comment by m2jobe** [comment]
Hi Alex thank you! I stumbled on that earlier but why are we creating for example table authorize_group there but the installer is asking for uf_authorize_group?

**Comment by alexweissman** [comment]
Oh!  It is probably grabbing the table prefix from your configuration file.  You can change that to `''` if you don't want prefixes (I actually recommend not having prefixes).  Can you join us in [chat](https://chat.userfrosting.com) to continue this discussion?

**Comment by m2jobe** [comment]
Ah I see thank you and yes I can!

**Comment by m2jobe** [comment]
Hi @xoorath what alternative did you use to authenticate your users? Like you i am working with an azure sql db.

**Comment by alexweissman** [comment]
@m2jobe I took a stab at implementing this in the `hotfix-development` branch.  It seems to work fine in MySQL - can you please pull and test with your database?  You can now set a value for `db_driver` in your `config-userfrosting.php`.  Allowed values are `sqlite`, `mysql`, `pgsql`, and `sqlsrv`.

**Comment by xoorath** [comment]
@m2jobe, I'm not using an alternative built on azure sql db. I pivoted entirely to another solution and am using nodejs with mongodb.

For what it's worth, User Frosting looked like a very nice solution. If they resolve this issue for the DB you're using, I'd see it through. Just my two cents.

**Comment by m2jobe** [comment]
@xoorath cheers for the response! 
And @alexweissman Ill test it out with my db and let you know if it works

**Comment by alexweissman** [comment]
Anyone want to test this so I can merge it in to `master`?

**Comment by Silic0nS0ldier** [comment]
I've had a crack at getting UserFrosting running with an Azure SQL Database. It was successfully able to connect (as indicated by the creation of the tables) but hit an exception.

Details
---
```
Type: PDOException
Code: 42000
Message: SQLSTATE[42000]: [Microsoft][ODBC Driver 11 for SQL Server][SQL Server]Incorrect syntax near '`'.
File: D:\xampp\app\htdocs\UserFrosting\userfrosting\models\database\SiteSettings.php
Line: 546
```

Trace
---
```
#0 D:\xampp\app\htdocs\UserFrosting\userfrosting\models\database\SiteSettings.php(546): PDOStatement->execute(Array)
#1 D:\xampp\app\htdocs\UserFrosting\userfrosting\models\database\UFModel.php(88): UserFrosting\SiteSettings->save()
#2 D:\xampp\app\htdocs\UserFrosting\userfrosting\models\database\Database.php(305): UserFrosting\UFModel->store()
#3 D:\xampp\app\htdocs\UserFrosting\userfrosting\controllers\InstallController.php(79): UserFrosting\Database::install()
#4 D:\xampp\app\htdocs\UserFrosting\public\index.php(328): UserFrosting\InstallController->pageSetupDB()
#5 [internal function]: {closure}()
#6 D:\xampp\app\htdocs\UserFrosting\userfrosting\vendor\slim\slim\Slim\Route.php(468): call_user_func_array(Object(Closure), Array)
#7 D:\xampp\app\htdocs\UserFrosting\userfrosting\vendor\slim\slim\Slim\Slim.php(1357): Slim\Route->dispatch()
#8 D:\xampp\app\htdocs\UserFrosting\userfrosting\vendor\slim\slim\Slim\Middleware\Flash.php(85): Slim\Slim->call()
#9 D:\xampp\app\htdocs\UserFrosting\userfrosting\vendor\slim\slim\Slim\Middleware\MethodOverride.php(92): Slim\Middleware\Flash->call()
#10 D:\xampp\app\htdocs\UserFrosting\userfrosting\middleware\CsrfGuard.php(51): Slim\Middleware\MethodOverride->call()
#11 D:\xampp\app\htdocs\UserFrosting\userfrosting\middleware\usersession\UserSession.php(27): Slim\Extras\Middleware\CsrfGuard->call()
#12 D:\xampp\app\htdocs\UserFrosting\userfrosting\vendor\slim\slim\Slim\Middleware\PrettyExceptions.php(67): UserFrosting\UserSession->call()
#13 D:\xampp\app\htdocs\UserFrosting\userfrosting\vendor\slim\slim\Slim\Slim.php(1302): Slim\Middleware\PrettyExceptions->call()
#14 D:\xampp\app\htdocs\UserFrosting\public\index.php(416): Slim\Slim->run()
#15 {main}
```

Issue looks to be tied to some remaining raw sql statements in the SiteSettings.php file.

**Comment by alexweissman** [comment]
Ah, thanks for catching that!  Let me see what I can do.

**Comment by alexweissman** [comment]
Alright, fixed - could you pull and test again?

**Comment by Silic0nS0ldier** [comment]
That seems to have solved that issue, and revealed an issue that I had overlooked while trying to get it to connect. (it only hits the error if tables are missing) Its crashing when seeding tables. (so around line 308 of userfrosting\models\database\Database.php).

Interestingly, the issue isn't actually caused by UserFrosting, but rather SQL Server (Azure and proper) blocking the specification of the identities. (http://stackoverflow.com/questions/1334012/cannot-insert-explicit-value-for-identity-column-in-table-table-when-identity)

Can still get around this by inserting without specifying the id, and then grabbing it from the database to use where needed later on. (or manually specifying the expected index, but that would just be asking for issues in my opinion)

**Comment by alexweissman** [comment]
Yeah, the other way we could do this is by actually building the Eloquent active record models, setting up the relationships, storing them, etc.  I just figured using the query builder was easier, but this requires us to explicitly state values for the `id` columns, so that we can reliably reference them in our intermediate tables.  For example, in the `authorize_group` table, we need to reference the `id` column in the `group` table.

**Comment by Silic0nS0ldier** [comment]
Okay, made some hacky modifications to the aforementioned file (`Database.php`) and the `InstallController`. (since it has to specify the `id` for the admin role). And its working. So Azure SQL Server (and regular SQL Server since Azure SQL is just a subset of it) can both work with UserFrosting. (just need to revise to not explicitly state `id`'s. presumably just the auto incrementing ones)

**Comment by alexweissman** [comment]
So, what changes did you make?  Just remove the `id` property from each insert?

**Comment by Silic0nS0ldier** [comment]
More or less, yeah. Specifics are below.

### Database.php Changes
https://github.com/Silic0nS0ldier/UserFrosting/commit/2a1ee4da890e98735beff5f2d3cf3adc8b16ad80

### InstallController.php Changes
https://github.com/Silic0nS0ldier/UserFrosting/commit/6ecdbbd282407e1403bbbc77cfc6e8a3562e1568

**Comment by Silic0nS0ldier** [comment]
Wait, one of those has old code.

**Comment by Silic0nS0ldier** [comment]
For the Database.php file, lines 338 to 340 are...
```
$user = Capsule::table(static::getSchemaTable('group')->name)->where('name', '=', 'User')->get()[0]['id'];
$admin = Capsule::table(static::getSchemaTable('group')->name)->where('name', '=', 'Administrator')->get()[0]['id'];
$zerglings = Capsule::table(static::getSchemaTable('group')->name)->where('name', '=', 'Zerglings')->get()[0]['id'];
```

**Comment by alexweissman** [comment]
Do you want to submit a PR to the `hotfix-development` branch?

Btw, you can use `->first()->id` instead of `->get()[0]['id']`.

**Comment by Silic0nS0ldier** [comment]
I did try `first`, but for whatever reason it didn't like it. (Can't remember the specific exception thrown) I'll give the code a clean, and submit a PR later today.

**Comment by alexweissman** [comment]
Yep!  But in Laravel, it is exposed via the `setBindings` method.  See http://stackoverflow.com/a/20873009/2970321

Actually, you probably don't even need that level of control - why not just use `insert`?  Then you could convert the user data to an array using `toArray()` and just pass it into the call to `insert`.

**Comment by alexweissman** [comment]
@m2jobe @xoorath, can you go ahead and test out the `hotfix-development` branch now?  If all goes well, we can go ahead and create a new release.

**Comment by alexweissman** [comment]
Implemented in [5f3cb53](https://github.com/userfrosting/UserFrosting/commit/5f3cb538ed4086f9afc71758c2e2ee53d533f7e3).
